tests/expose_test.go

```diff
@@ -507,50 +508,86 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 
 	Context("Expose a VM as a service.", func() {
 		const servicePort = "27017"
-		const serviceName = "cluster-ip-vm"
+		const serviceNamePrefix = "cluster-ip-vm"
 		var vm *v1.VirtualMachine
 
-		tests.BeforeAll(func() {
-			tests.BeforeTestCleanup()
-			By("Creating an VM object")
-			template := newLabeledVMI("vm", virtClient, false)
-			vm = tests.NewRandomVirtualMachine(template, false)
-
-			By("Creating the VM")
-			_, err := virtClient.VirtualMachine(tests.NamespaceTestDefault).Create(vm)
-			Expect(err).ToNot(HaveOccurred())
-
-			By("Exposing a service to the VM using virtctl")
-			virtctl := tests.NewRepeatableVirtctlCommand(expose.COMMAND_EXPOSE, "virtualmachine", "--namespace",
-				vm.Namespace, vm.Name, "--port", servicePort, "--name", serviceName, "--target-port", strconv.Itoa(testPort))
-			err = virtctl()
-			Expect(err).ToNot(HaveOccurred())
+		var vmExposeArgs []string
 
+		startVMIFromVMTemplate := func(name string, namespace string) *v1.VirtualMachineInstance {
 			By("Calling the start command")
-			virtctl = tests.NewRepeatableVirtctlCommand("start", "--namespace", vm.Namespace, vm.Name)
-			err = virtctl()
-			Expect(err).ToNot(HaveOccurred())
+			virtctl := tests.NewRepeatableVirtctlCommand("start", "--namespace", namespace, name)
+			Expect(virtctl()).To(Succeed(), "should succeed starting a VMI via `virtctl start ...`")
 
 			By("Getting the status of the VMI")
 			Eventually(func() bool {
-				vm, err = virtClient.VirtualMachine(vm.Namespace).Get(vm.Name, &k8smetav1.GetOptions{})
+				vm, err = virtClient.VirtualMachine(namespace).Get(name, &k8smetav1.GetOptions{})
 				Expect(err).ToNot(HaveOccurred())
 				return vm.Status.Ready
 			}, 120*time.Second, 1*time.Second).Should(BeTrue())
 
 			By("Getting the running VMI")
 			var vmi *v1.VirtualMachineInstance
 			Eventually(func() bool {
-				vmi, err = virtClient.VirtualMachineInstance(vm.Namespace).Get(vm.Name, &k8smetav1.GetOptions{})
+				vmi, err = virtClient.VirtualMachineInstance(namespace).Get(name, &k8smetav1.GetOptions{})
 				Expect(err).ToNot(HaveOccurred())
 				return vmi.Status.Phase == v1.Running
 			}, 120*time.Second, 1*time.Second).Should(BeTrue())
 
-			tests.GenerateHelloWorldServer(vmi, testPort, "tcp")
-		})
+			return vmi
+		}
+
+		createStoppedVM := func(virtClient kubecli.KubevirtClient, namespace string) (*v1.VirtualMachine, error) {
```

Comment:
Some of the functions use the outer `virtClient` (startVMIFromVMTemplate), some do not.
In the same context, why you pass in the namespace? I do not see now anything else using it with a different input.

I guess this is a repeated pattern here. If the lambda is not either inlined or using outer values to construct it, I find it very confusing (and requires from my part to lookup for which vars you do use).

Reply:
I understand that lambdas that use stuff defined out of their body is extremely confusing, and reading that code without an IDE a nightmare - that's why I make a serious effort of passing the attributes to the functions. Please assume all variables must be passed as arguments, and keep letting me know when they don't.

Reply:
Well, that rule is just contradicting the existence of lambdas in Go IMO.
I do not think they exist to limit scope, they exist to either do something inlinned or use an outer variable in order to carry a dynamic state (i.e. it is a generated function and not a static one).